Prompt API: Add overview for Prompt API (#43848)
mdn/content: files/en-us/web/api/prompt_api/using/index.md

---
title: Using the Prompt API
slug: Web/API/Prompt_API/Using
page-type: guide
---

{{DefaultAPISidebar("Prompt API")}}

> The [Prompt API](/en-US/docs/Web/API/Translator_and_Language_Detector_APIs) provides an asynchronous ({{jsxref("Promise")}}-based) mechanism for a website to directly prompt a language model provided by the user agent, without needing to manage implementation-specific details of the AI model being used. This is useful and efficient because sensitive data can stay on the user's device, the model is available offline, and developers can avoid the cost and latency of API calls to external services.

Reviewer: "This is useful" is a bit odd because I read "this" as "without needing to manage implementation-specific details of the AI model being used." but the rest refers to the benefits of an on-device model.

I think I'd say "Having an on-device model is" rather than "This".

Reviewer 2: Fair; updated.

This article explains how to use the core fundamentals of the Prompt API.

## Creating a session

> All of the AI prompting functionality is contained within the {{domxref("LanguageModel")}} interface.

Reviewer: Maybe

```suggestion
All of the AI prompting functionality is managed via the {{domxref("LanguageModel")}} interface.
```

Reviewer 2: Like it. Updated.

The Prompt API handles text inputs and outputs by default, but it is multimodal — you can also give it images and audio inputs, for example to ask it to describe an image or transcribe an audio file. See [Multimodal prompts](/en-US/docs/Web/API/Prompt_API/Multimodal) for more details.

The Prompt API will handle multiple languages by default, but it might not handle all languages you are expecting, so it is a good idea to explicitly specify them in case the browser needs to download extra resources.

## Prompting the model

When you've created a `LanguageModel` instance, you can start prompting the AI model by calling the {{domxref("LanguageModel.prompt()")}} instance method on it, passing it one or more input prompts as an argument. For example:

```js
const response = await session.prompt(textarea.value);
```

This method returns a {{jsxref("Promise")}} that fulfills with a string containing the AI response to your prompt.

> You can pass multiple inputs into the API with different roles — for example, standard `user` prompts and instructions from the `assistant` to further shape how it responds to the `user` prompts. To get the AI to respond to your input in the style of a villanous mastermind, you might use this `prompt()` call:

Reviewer: Are the instructions from or to the assistant? The text here an example don't quite match.

What I mean is in the following example you have a suggestion for the assistant: 

>    content: "I'm going to answer this user query like a James Bond villain",

I read that as something the AI reads back verbatim to the user. That would reflect the case where you are showing some answer/responses for the AI to copy. But the text implies you're just passing instructions to the AI about how it should response, in which case I would expect the example to be:

>    content: "Answer the user as a James Bond villain",

This really matters. My biggest gripe with the original version was that it wasn't clear what the user roles were for. It isn't clear what "few-shot" and "continued dialogue" are. These terms are very AI centric and need clear examples and definitions.

This for example is completely unhelpful: https://pr43848.review.mdn.allizom.net/en-US/docs/Web/API/LanguageModel/create_static#assistant

> ["assistant"](https://pr43848.review.mdn.allizom.net/en-US/docs/Web/API/LanguageModel/create_static#assistant)
A message from the model. Use this for few-shot examples or continued dialogue. A few-shot example is a set of input-output pairs passed as an example to an AI before asking it to complete a similar task.

- `A message from the model. ` - how is it **From** the model? You're seeding the model with it.
- What is a few-shot, and how do I use this feature for that.
- What is a continued dialog (NOTE SPELLING - do we use dialog or dialogue?) and how does this relate.

Yes we hint at this " A few-shot example is a set of input-output pairs passed as an example to an AI before asking it to complete a similar task." but it is core and needs to be in the guide.

Reviewer: I see some of this is covered in the context guide. Should link to that from above.

Reviewer 2: I've updated the reference pages to remove this jargon and make the description of the role more consistent with what I've written in the guides.

I've also updated the James Bond villain prompt to your suggestion.

The "few-short" terminology is mentioned in one of the reference examples still, in the `create()` page. I've left that for now. I can add a quick bit about few-shot to the guide somewhere if you think I need it, but I deliberately kept this fairly light in terms of prompt engineering tuition, in the interests of trying to get it published fairly promptly (pun intended).

Reviewer: >  get it published fairly promptly (pun intended).

I will never forgive you for this :-).

Reviewer 2: I'm sure you will in time. You are not as much of a curmudgeon as your reviews suggest ;-)

Reviewer: Shhhhhhh. Don't ruin my rep.

```js
const response = await session.prompt([
  {
    role: "assistant",
    content: "I'm going to answer this user query like a James Bond villain",
  },
  {
    role: "user",
    content: textarea.value,
  },
]);
```

You'll learn more about these roles in the next article, [Adding context with initial and ongoing prompt inputs](/en-US/docs/Web/API/Prompt_API/Adding_context).

> [!NOTE]
> If you want to return the AI response gradually as a {{domxref("ReadableStream")}} rather than a single large string, you can use the {{domxref("LanguageModel.promptStreaming()")}} method.

> ## Checking configuration support

Reviewer: I think we need to explain why this is useful. Why not just start a session and have it fail? 

If there is a cost to trying and we should do this, then I'd argue that this section should appear first. If there is no cost it is in the right place, but we should still explain why you might bother.

Reviewer 2: Agreed. I've rejigged the order and wording so that the config check section comes first. I've also added a bit more description as to why you should do this.

## Monitoring download progress

If the AI model is downloading (`availability()` returns `downloading`), it is helpful to provide the user with feedback to tell them how long they need to wait before the operation completes.

The `create()` method can accept a `monitor` property, the value of which is a callback function that takes a {{domxref("CreateMonitor")}} instance as an argument. `CreateMonitor` has a {{domxref("CreateMonitor/downloadprogress_event", "downloadprogress")}} event available, which fires when progress is made on downloading the AI model.

You can use this event to expose loading progress data:

```js
const session = await LanguageModel.create({
  expectedInputs: [{ type: "text", languages: ["en"] }],
  expectedOutputs: [{ type: "text", languages: ["en"] }],
  monitor(monitor) {
    monitor.addEventListener("downloadprogress", (e) => {
      promptOutput.textContent = `Downloading model data ${Math.floor(e.loaded * 100)}%`;
    });
  },
});
```

If the specified languages are not supported, a download will not be initiated, and a `NotSupportedError` {{domxref("DOMException")}} will be thrown.

## The context window

> Every `LanguageModel` session has a finite context window, which constrains the total number of input and output tokens it can hold at once. The {{domxref("LanguageModel.contextWindow", "contextWindow")}} property reports the session's maximum capacity, and {{domxref("LanguageModel.contextUsage", "contextUsage")}} reports how many tokens have been consumed so far.

Reviewer: I think of tokens as a finite resource I pay for (ie. Claude). When I run out I have to wait until I pay for more or my usage resets. I understand here we're talking more about a capacity of a session right - so it runs out not because of payment but because of physical resource limitations?

If so, it might be worth making an explicit note about that false assumption. More important, also capture the things you can do to stop running out of resource!

Reviewer 2: I have added a bit more description around this. I am not sure exactly what decides each session's token allowance, and whether you can get more tokens by just sitting and waiting or whether that's it and you have to clone the session. 

Maybe @tomayac can offer some thoughts here, although I'm not sure how much it is worth saying here.

Reviewer: I do think it would be useful to understand how tokens are consumed and map to resources. 

This is not so much because I want to understand how you can conserve resources, though that would be really useful.

The main reason is that as it stands now cloning is presented as a way to preserve tokens and continue your interaction. I suspect that isn't the purpose at all: my guess would be that the intent of cloning is to allow diverging interactions with the AI based on a common starting point, and the fact that you get a clean session is a happy circumstance, _not_ the point of the feature.

Knowing the "point" is what is important here, and at the moment we can only infer it.

Reviewer 2: I think it is a good idea to add some further info about using cloning to allow diverging interactions with the AI based on a common starting point. I'll update the section to provide that.

We'll have to wait on @tomayac to see if he has any further info about how token allowance is decided, how the tokens are consumed, and how they map to resources.

For example, after each prompt, you can report how many tokens are left using something like this:

```js
console.log(`${session.contextUsage}/${session.contextWindow}`);
```

When a method call such as {{domxref("LanguageModel.prompt()", "prompt()")}} or {{domxref("LanguageModel.promptStreaming()", "promptStreaming()")}} would exceed the remaining number of tokens in the context window, a `QuotaExceededError` {{domxref("DOMException")}} is thrown and the {{domxref("LanguageModel.contextoverflow_event", "contextoverflow")}} event fires.

To check how many tokens a prompt operation would consume without actually sending it, use {{domxref("LanguageModel.measureContextUsage()", "measureContextUsage()")}}.

## Cloning a session

> One way to get around the problem of running out of tokens is to copy your existing session using the {{domxref("LanguageModel.clone()")}} function. This creates a replica of the `LanguageModel` object instance in which the conversation up to that point and initial prompt are preserved, but the token count (`contextUsage`) is reset. You can think of the session clone as being a fork of the original conversation, with its own token allowance.

Reviewer: See my other comment. I think it unlikely that cloning only exists to reset tokens. If that were the case you'd just have a resetTokens() method or similar.

Of course I could be wrong, in which case we should be very clear that is its only purpose.

Reviewer 2: I've rewritten this section to add a diverging paths example and make the focus more about that. I still mention the solution to running out of tokens as another use case.

Reviewer: The new text is great.

You can cancel pending `prompt()`, `clone()` and other operations using an {{domxref("AbortController")}}, with the associated {{domxref("AbortSignal")}} being included inside the method options object as a `signal` property value. For example, aborting a `LanguageModel.prompt()` operation via a button press could look like this:

```js
const controller = new AbortController();

abortBtn.addEventListener("click", () => {
  controller.abort("Query aborted by user.");
});

const response = await session.prompt(textarea.value, {
  signal: controller.signal,
});
```

After a `LanguageModel` has been created, you can release its assigned resources and stop any further activity by calling its {{domxref("LanguageModel.destroy()")}} method. You are encouraged to do this after you've finished with the object as it can consume a lot of resources.

```js
session.destroy();
```

If a `create()` call has an associated {{domxref("AbortController")}}, and you call its {{domxref("AbortController.abort()")}} method after the `create()` call has succeeded, it will have the same effect as calling `destroy()` on the resulting `LanguageModel` object.

## Complete example

> Let's look at a complete basic example that demonstrates the Prompt API in action.

Reviewer: Should we say what it delivers as well. - "The example xxx"?

Reviewer 2: Good idea. Added.

### HTML

> In our markup, we define an input {{htmlelement("textarea")}} that allows the user to type in a prompt. We also include two {{htmlelement("button")}} elements — one to submit the prompt/query, and another to abort an ongoing query.

Reviewer: FMI, do we have to have any concern about the user input? I mean can we assume that the AI will sanitize any input. Similarly, do we need to do anything with the output before rendering or do we assume it is safe?

Reviewer 2: Do you mean, security-wise, or harmful content-wise? If the former, well, I'm assuming Google sanitises content sent into their browser internals. If the latter, I tried a few tests, and the AI won't return anything harmful (e.g. hateful, violent, or sexually explicit).

In terms of output, it seems to always output markdown, which you could argue should be converted to HTML before being injected into the page...I was kinda hoping to ignore that for now in the intitial docs phase.

Reviewer 2: Correction: it is always a string, which can be markdown or a JSON string, depending on what you ask it to output. I still don't really want to go into handling outputs in lots of detail, at least not initially.

Reviewer: I'm OK with that for now. This was really just something that popped into my head.

Note though, I think we should treat an AI like any other untrusted input: we don't manage its content or know what it was trained on. It might be Skynet.

Reviewer 2: True, true.

- `unavailable` means that the implementation doesn't support the requested options.

If a download is required, it will be started automatically by the browser once a `LanguageModel` instance is created using the `create()` method. You can track download progress automatically using a monitor, which we'll cover in the next section.

> [!NOTE]
> Even though you can ask for a language model session that expects multimedia outputs, this will fail — the availability will be `unavailable`. The API currently only supports text outputs.

### Monitoring download progress

If the AI model is downloading additional data (`availability()` returns `downloading`), it is helpful to provide the user with feedback to tell them how long they need to wait before the operation completes.

The `create()` method can accept a `monitor` property, the value of which is a callback function that takes a {{domxref("CreateMonitor")}} instance as an argument. `CreateMonitor` has a {{domxref("CreateMonitor.downloadprogress_event", "downloadprogress")}} event available, which fires when progress is made on downloading the data.

> You can use this event to expose loading progress data:

Reviewer: ```suggestion
You can use this event to get the loading progress:
```

```js
const session = await LanguageModel.create({
  expectedInputs: [{ type: "text", languages: ["en"] }],
  expectedOutputs: [{ type: "text", languages: ["en"] }],
  monitor(monitor) {
    monitor.addEventListener("downloadprogress", (e) => {
      promptOutput.textContent = `Downloading model data ${Math.floor(e.loaded * 100)}%`;
    });
  },
});
```

> If the specified languages are not supported, a download will not be initiated, and a `NotSupportedError` {{domxref("DOMException")}} will be thrown.

Reviewer: No the returned promise will reject with the error - it isn't thrown ... unless you're assuming all code here is using try/catch/await semantics?

Reviewer: Ignore, in an await it is thrown.

## Creating a `LanguageModel` session

Once you have checked that your configuration is supported, the next step in prompting the AI model is to create a `LanguageModel` object instance. This is done using the {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} static method, which takes an options object as an argument:

```js
const session = await LanguageModel.create({
  expectedInputs: [{ type: "text", languages: ["en"] }],
  expectedOutputs: [{ type: "text", languages: ["en"] }],
});
```

The browser will automatically download the corresponding model data to handle the requested language model if it is not already available, and if the browser is able to do so.

> [!NOTE]

> > The `create()` method (and other methods available via the Prompt API) require [transient activation](/en-US/docs/Glossary/Transient_activation) to invoke, as a precaution to stop apps from using language model resources without user interaction.

Reviewer: Maybe

```suggestion
> The `create()` method (and other methods available via the Prompt API) require [transient activation](/en-US/docs/Glossary/Transient_activation) to invoke, so that apps can only use language model resources following user interaction.
```

A `LanguageModel` object instance and the activity that occurs as a result of using its methods and properties is called a **session**. The browser stores all the prompts and responses sent to and received from the Prompt API as part of a single session, allowing the API to tailor its responses based on previous interactions and hold a conversation.

This includes any prompt messages sent to it via the {{domxref("LanguageModel.create_static", "create()")}} method's `initialPrompts` option, {{domxref("LanguageModel.prompt", "prompt()")}}, {{domxref("LanguageModel.promptStreaming", "promptStreaming()")}}, or {{domxref("LanguageModel.append", "append()")}}.

> [!NOTE]

> > The browser doesn't store session information across browser reloads by default. To restore session context after a reload or browser restart, you will have to implement a mechanism to save the conversation and restore it using a server-side solution or a client-side mechanism such as [Web Storage](/en-US/docs/Web/API/Web_Storage_API). Such an example is covered in [Preserving sessions across reloads](/en-US/docs/Web/API/Prompt_API/Preserving_sessions).

Reviewer: Not wrong, but "Such an" felt a little clunky
```suggestion
> The browser doesn't store session information across browser reloads by default. To restore session context after a reload or browser restart, you will have to implement a mechanism to save the conversation and restore it using a server-side solution or a client-side mechanism such as [Web Storage](/en-US/docs/Web/API/Web_Storage_API). An example is provided in [Preserving sessions across reloads](/en-US/docs/Web/API/Prompt_API/Preserving_sessions).
```

The [`expectedInputs`](/en-US/docs/Web/API/LanguageModel/create_static#expectedinputs) and [`expectedOutputs`](/en-US/docs/Web/API/LanguageModel/create_static#expectedOutputs) parameters specify the types of input and output and the input/output languages you are expecting to provide to and receive from the AI prompt.

The Prompt API handles text inputs and outputs by default, but it is multimodal — you can also give it images and audio inputs, for example to ask it to describe an image or transcribe an audio file. See [Multimodal prompts](/en-US/docs/Web/API/Prompt_API/Multimodal) for more details.

The Prompt API will handle multiple languages by default, but it might not handle all languages you are expecting, so it is a good idea to explicitly specify them in case the browser needs to download extra resources.

## Prompting the model

When you've created a `LanguageModel` instance, you can start prompting the AI model by calling the {{domxref("LanguageModel.prompt()")}} instance method on it, passing it one or more input prompts as an argument. For example:

```js

> const response = await session.prompt(textarea.value);

Reviewer: Consider using a literal - takes away questions around trusting the input.

```suggestion
const response = await session.prompt("Why does a fish need a bicycle?");
```

```

This method returns a {{jsxref("Promise")}} that fulfills with a string containing the AI response to your prompt.

> You can pass multiple inputs into the API with different roles — for example, standard `user` prompts and instructions from the `assistant` to further shape how it responds to the `user` prompts. To get the AI to respond to your input in the style of a villainous mastermind, you might use this `prompt()` call:

Reviewer: ```suggestion
You can pass multiple inputs into the API with different roles — for example, standard `user` prompts and instructions for the `assistant` to further shape how it responds to the `user` prompts. To get the AI to respond to your input in the style of a villainous mastermind, you might use this `prompt()` call:
```

  },
  {
    role: "user",
    content: textarea.value,
  },
]);
```

You'll learn more about these roles in the next article, [Adding context with initial and ongoing prompt inputs](/en-US/docs/Web/API/Prompt_API/Adding_context).

> [!NOTE]
> If you want to return the AI response gradually as a {{domxref("ReadableStream")}} rather than a single large string, you can use the {{domxref("LanguageModel.promptStreaming()")}} method.

## The context window

Every `LanguageModel` session has a finite context window, which constrains the total number of input and output tokens it can hold at once. Once you use up your session's token allowance, you cannot issue any more prompts, and you need to use a technique such as [session cloning](#cloning_a_session) to continue usage.

The {{domxref("LanguageModel.contextWindow", "contextWindow")}} property reports the session's maximum capacity, and {{domxref("LanguageModel.contextUsage", "contextUsage")}} reports how many tokens have been consumed so far.

For example, after each prompt, you can report how many tokens are left using something like this:

```js
console.log(`${session.contextUsage}/${session.contextWindow}`);
```

> When a method call such as {{domxref("LanguageModel.prompt()", "prompt()")}} or {{domxref("LanguageModel.promptStreaming()", "promptStreaming()")}} would exceed the remaining number of tokens in the context window, a `QuotaExceededError` {{domxref("DOMException")}} is thrown and the {{domxref("LanguageModel.contextoverflow_event", "contextoverflow")}} event fires.

Reviewer: Another case of "is thrown" - errors are rejected promises. Can you do a search/replace check?

    },
  ],
});

// ...

// Science quiz clone
const firstClone = await session.clone();
await firstClone.prompt("Give me a question about science.");
await firstClone.prompt("Another question, please.");

// 80's music quiz clone
const secondClone = await session.clone();
await secondClone.prompt("Give me a question about 80's popular music.");
await secondClone.prompt("Another question, please.");
```

Creating a new session via `clone()` is also a common way to get around the problem of running out of tokens.

## Cancelling operations and destroying instances

You can cancel pending `prompt()`, `clone()` and other operations using an {{domxref("AbortController")}}, with the associated {{domxref("AbortSignal")}} being included inside the method options object as a `signal` property value. For example, aborting a `LanguageModel.prompt()` operation via a button press could look like this:

```js
const controller = new AbortController();

abortBtn.addEventListener("click", () => {
  controller.abort("Query aborted by user.");
});

const response = await session.prompt(textarea.value, {
  signal: controller.signal,
});
```

> After a `LanguageModel` has been created, you can release its assigned resources and stop any further activity by calling its {{domxref("LanguageModel.destroy()")}} method. You are encouraged to do this after you've finished with the object as it can consume a lot of resources.

Reviewer: Pedantic, but you can't destroy what doesn't exist. Something like

```suggestion
When you have finished with a {{domxref("LanguageModel")}} you can release its assigned resources and stop any further activity by calling its {{domxref("LanguageModel.destroy()")}} method. You are encouraged to do this when you finish your session, as it can consume a lot of resources.
```

* {
  box-sizing: border-box;
}

html {
  font-family: Arial, Helvetica, sans-serif;
}

body {
  max-width: 600px;
  margin: 0 auto;
}

form div {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 20px;
}

textarea,
.prompt-output {
  padding: 5px;
}

.prompt-output {
  min-height: 150px;
  border: 1px solid black;
  width: 100%;
  display: block;
}

.error {
  color: red;
}

button {
  margin-right: 10px;
}
```

Note that we won't show the CSS for this example, as none of it is relevant to understanding the Prompt API.

### JavaScript

In our script, we start off by grabbing references to the `<form>`, `<textarea>`, submit `<button>`, abort `<button>`, and output `<p>`. We initially disable the submit and abort buttons, as we don't want them to be pressed before the related functionality is available.

```js live-sample___prompt-example
const form = document.querySelector("form");
const textarea = document.querySelector("textarea");
const submitBtn = document.querySelector("#submit");
const abortBtn = document.querySelector("#abort");
abortBtn.disabled = true;
submitBtn.disabled = true;
const promptOutput = document.querySelector(".prompt-output");
```

> Next, we create a global `session` variable to hold our session. Because using the API requires transient activation, we populate `session` inside a `focus` event handler on the `<textarea>`. When the user focuses the `<textarea>`, we first check whether the API is supported; if not, we print a non-support message and `return` early. Next, we check whether `session` already has a value assigned (we don't want to create a new session each time). If not, we run the `init()` function, which generates a `LanguageModel` instance using the custom `getSession()` function defined later on.

Reviewer: FWIW You're checking API support based on LanguageModel being defined on the window. That doesn't need transient activation.

This matters a little, because my first thought was "Oh, you have to interact to display support, that's a flaw in the API design"
